Add UTF encoding validity functions

ScottPJones opened the pull request:
This introduces the `Base.check_string` function, with methods for handling UTF-8, UTF-16, and UTF-32 as vectors of `UInt8`, `UInt16`, and `UInt32`, respectively, as well as an `AbstractString` version that operates on Unicode characters (making sure that they are all valid code points, `0 <= ch < 0xd800`, `0xe000 <= ch < 0x10ffff`).
There are options to accept or not accept things like `Modified UTF-8` encoding, or `CESU-8` encoding,
or "overly long" encodings.  (See `utfcheck.jl` for documentation)
These methods either throw a `UnicodeError`, with information about which character was invalid, and
its position in the input, or they return a tuple, of the number of logical characters in the string, a bit flag to indicate what types of data were found (i.e. all ASCII, all Latin1, surrogates present, overlong characters present, etc.), the number of characters that would take 2 bytes to encode in UTF-8, the number that would take 3 bytes, and the number that would take 4 bytes.
This information is designed to be useful for future conversion routines, because they allow one to calculate the exact size needed to represent a valid string in UTF-8, UTF-16, or UTF-32, and also
indicate whether an optimized "widening" or "narrowing" conversion can be done.

ScottPJones committed:
Add Unicode validation function
    Adds `check_string` function, which checks a vector of bytes,
    16-bit or 32-bit words, or an AbstractString for validity,
    either for UTF-8, UTF-16, or UTF-32 encoding.
    By default, `Modified UTF-8 (long \0 encoding)` and
    `CESU-8 (surrogate pairs encoded as 2 UTF-8 3-byte sequences)`
    are allowed, but other over long encoded sequences are not allowed,
    but this can be changed by the keyword options argument.
Add unit tests of all the errors found by `check_string`
Updated documentation to not use doxygen tags.

Move documentation strings from line after to line before

Add testing of valid strings
Improve/consolidate documentation

Add bounds checking

Change name to unsafe_checkstring, warn that doesn't check bounds
Add checkstring, which does check bounds

Add tests of bounds checking

Change order of start/end positions

Update bounds checking tests

Change 1 to start(dat)

Use checkbounds()

ScottPJones commented:
Prepare your :tomato:s!  I'm off for dinner, please let me know what you think of this and #11573

tkelman commented on base/sysimg.jl:
```diff
@@ -84,6 +84,8 @@ include("iterator.jl")
 
 # strings & printing
+include("utferror.jl")
+include("utftypes.jl")
```
you'll need to `include("utfcheck.jl")` here too

ScottPJones replied:
Hmm... somehow that got lost... was definitely part of my source locally...

vtjnash commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,173 @@
+function check_string{T <: Union(Vector{UInt16}, Vector{UInt32}, AbstractString)}(
+                      dat::T,
+                      len::Int = (T == AbstractString ? endof(dat) : (sizeof(dat)>>>(T==Vector{UInt16}?1:2))),
+                      pos::Int = (T == AbstractString ? start(dat) : 0)
```
i believe this should work with `start(dat)` always. note that `start(dat)` isn't an `Int` in the general case however

ScottPJones replied:
I'll try that out, look at the generated code when I get home... thanks!

nalimilan commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,173 @@
+                elseif is_surrogate_codeunit(ch)
+                    throw(UnicodeError(UTF_ERR_SURROGATE, pos-3, ch))
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    # This is an overly long encode character
```
"encoded"?

ScottPJones replied:
yep, mr. eagle eyes!  thanks!

nalimilan commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,173 @@
+
+## Options for check_string_* functions
+
+const UTF_NO_LONG_NULL = 1      ##< don't accept 0xc0 0x80 for '\0'
```
What's this `##<` sequence?

ScottPJones replied:
Doxygen markup... makes the comments be added to the enum, structure, or constant definition documentation.

nalimilan replied:
Ah, OK.

hayd replied:
cc @MichaelHatherly I wonder if this (or something similar) c/should be something Docile does?

tkelman commented:
Please rebase relative to latest master now that #11573 is merged

ScottPJones commented:
Of course!  Thanks very much.  Should I squash everything also?

tkelman commented:
It's only 2 commits here, and rebasing would remove 1 since that's now on master.

If you want to amend your commit message while rebasing to make it a little more descriptive that would be welcomed.

tkelman commented:
As I mentioned at https://github.com/JuliaLang/julia/pull/11551#issuecomment-108800765, I think this code is essentially ready and good to merge, assuming the tests pass.

ScottPJones commented:
Is that what you wanted, as far as the commit message?  Thanks!

tkelman commented:
The indentation is a little odd, but it's fine. Nice and descriptive, I like that part.

ScottPJones commented:
Ah, I see about what happened to the indentation... I updated that directly on GitHub.
The test failure is the dreaded OOM, not the fault of this change, so hopefully somebody can merge this.
Thanks!

vtjnash commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,187 @@
+            elseif ch < 0x800
+                num2byte += 1
+                flags |= UTF_UNICODE2
+            elseif T != Vector{UInt16} && ch > 0x0ffff
```
is the type check here actually necessary? if T==UInt16, then typemax(T) == 0xffff && ch > 0xffff should be false. no?

ScottPJones replied:
It's checking against `Vector{UInt16}`, not `UInt16`, so I don't think Julia can figure it out by itself.

vtjnash replied:
```julia
julia> function t16(x)
         local a::UInt32
         @inbounds a, i = next(x, 1)
         if a > 0xffff
           return 1
         else
           return 2
         end
       end
t16 (generic function with 1 method)

julia> code_native(t16,Tuple{Vector{UInt16}})
	.section	__TEXT,__text,regular,pure_instructions
Filename: none
Source line: 4
	pushq	%rbp
	movq	%rsp, %rbp
	movl	$2, %eax
Source line: 4
	popq	%rbp
	ret

julia> code_llvm(t16,Tuple{Vector{UInt16}})

define i64 @julia_t16_20941(%jl_value_t*) {
top:
  ret i64 2
}
```
that looks pretty successful to me.

ScottPJones replied:
Thanks!  I changed that.  I know that at one point (before Jeff put in the change that fixed the big performance regression compared to 0.3.x I'd seen) it seemed not to get that correctly... (unless my eyes where playing tricks on me!), but I hadn't thought to try again (also my test wasn't the same as yours, it was using an AbstractString)

ScottPJones commented:
The only failures are due to the OOM problems affecting everybody.
I believe I've addressed the remaining issues.  Anything else preventing this from going in?
Thanks!

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,195 @@
+end
+
+function check_string(dat::Vector{UInt8}, len = sizeof(dat), pos = 0 ; options::Integer=0)
+" Validates and calculates number of characters in a UTF-8 encoded vector of UInt8
```
I believe @nalimilan said this in one of the other PR's, but the doc string convention is more likely to be docs immediately preceding the function, not just inside the function. Not positive though.

ScottPJones replied:
I hadn't seen that, but, why introduce a different convention from what's very commonly in use (i.e. Python documentation)? I'd heard that void strings (which I took to mean any string inside or outside of a method, since it wasn't at all specified, that wasn't used (as an argument, to be returned, or set into a variable), would in the future be picked out for documentation.
Since it seems there is a lot of overlap between Python and Julia programmers (all the SciPy, NumPy, etc. people), I think it would be a very good think to be consistent with that, if there is no compelling reason not to.

tkelman replied:
Compelling reason being Julia has multiple dispatch, with multiple method implementations for the same generic function but different argument types, which Python doesn't have. That and the `@doc` macro works for docstring-before-function, so I believe that's the direction that things are going.

ScottPJones replied:
The use of the `@doc` macro apparently will no longer be necessary (or even the `doc` string prefix, if the latest version of Docile were put in Base.  I'm not sure why multiple dispatch feels like a compelling reason to you... This doesn't remove the ability to (later in the build process, at least) use `@doc doc"""...""" ->` syntax before the first function of a set of methods...
I think you are going to see this style whether you like it or not... just this week I had to tell a Python programmer learning Julia that he had to move his comments out of the function and use `@doc doc"""...""" ->` in order to get Help to pick it up.  Allowing this style will make a lot of people moving from Python happy (or less confused, at least)

nalimilan replied:
People coming from Doxygen, Javadoc or gtk-doc are used to having docs before the body of the function, so you can't please everybody. Anyway, didn't you support a solution similar to Doxygen? :-p

ScottPJones replied:
doxygen doesn't have anything to do with it... you can use doxygen with Python, which has the docstring inside the function.
Also, I'm not saying that the default shouldn't be to have the documentation before... but why couldn't both be allowed?
About doxygen - I just like having things documented, and using doxygen pushes that a bit, but I've never been particularly attached to the *syntax*, just to having the necessary information preserved to make cross-references, etc.   Since nobody here seems to like the doxygen syntax, and somebody on another thread talked about parsing the comments to build the metadata, (which could then be used to produce output for doxygen), I decided that that was the better solution for now.

tkelman replied:
Can we just pick one convention, and use it without arguing please? Docstrings still need some work, but for code outside of base, docstring before function has been implemented, docstring inside of function hasn't. If someone wants to make a PR to support docstrings inside of functions in Docile or the base docs system, then that convention could be an alternate to accommodate people coming from python. For now, wouldn't it be easier to just make this change than waste more bandwidth resisting trivial changes that multiple reviewers are asking you to make?

ScottPJones replied:
I'm fine with changing it... I was responding via my phone until a couple of hours ago... out enjoying the day... I'm back at the laptop, doing it now.

ScottPJones commented:
I believe I've handled everything... please take a look!

tkelman commented:
@nalimilan @hayd do the docstrings look okay here now? This looks more or less good enough now to be worth merging, but I'd also like someone else with unicode experience to look over it carefully for correctness, since the majority of the style comments have been addressed now.

nalimilan commented:
Docstrings look OK to me (but I'm certainly not the best one to comment about what's the recommended style).

MichaelHatherly commented:
Here's how the first docstring is rendering with Docile:

![docs](https://cloud.githubusercontent.com/assets/6144086/8039091/c8806172-0e07-11e5-9d99-0b5737f79665.png)

Note the nested list isn't being handled correctly. Is there an open issure for that, @hayd?

The ``str`` argument should be called ``dat`` to match up with the signature shouldn't it?

Might be nice to wrap the ``str``, ``len``, ``pos``, ``options``, and ``UnicodeError`` in backticks to have them rendered as code rather than plain strings.

ScottPJones commented:
Thanks... I'll make those changes.

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,195 @@
+function check_string{T <: Union(Vector{UInt16}, Vector{UInt32}, AbstractString)}(
+                      dat::T,
+                      len = endof(dat),
+                      pos = (T <: AbstractString) ? start(dat) : 1
```
wasn't this formerly always using `start(dat)` for every case?

ScottPJones replied:
I had it this way at first, had tried just doing start(dat), ran into a strange error, and backed that out... but it might not have been from that line after all... I'll see if it works again with just start(dat).

quinnj replied:
What's the case where `1` doesn't work?

tkelman replied:
some subtypes of AbstractString I believe, reverse or substrings?

nalimilan replied:
I don't think such a case exists. When exploring https://github.com/JuliaLang/julia/issues/9297 I even considered allowing `1` as the only accepted integer to index a string, because it was painful to change all code to use `start(s)`.

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,195 @@
+" Validates and calculates number of characters in a UTF-8 encoded vector of UInt8
+
+### Input Arguments:
+* str    Vector of UInt8
```
variable name does not match actual signature

hayd commented:
(@MichaelHatherly nested MD blocks is on the radar / there's a TODO, not sure if there's an issue. IIRC @one-more-minute has a solution in mind.)

ScottPJones commented:
Another update... have at it!

tkelman commented on test/strings.jl:
```diff
@@ -1719,3 +1719,102 @@ d = UTF32String(c)
 
+# Test invalid sequences
+
+byt = 0x0
```
is this line needed? and unnecessary indentation of the chunk of tests below this

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,208 @@
+* `pos`    start position (defaults to `start(dat)`)
+
+### Keyword Arguments:
+* `accept_long_nul`   = `true` # Modified UTF-8 (`\\0` represented as `b\"\\xc0\\x80\"`)
```
The keyword argument is spelled `accept_long_null` in the signature of the other version, but you're spelling it `accept_long_nul` in the docstrings and the UTF8 method's signature. I'd prefer using the `null` version consistently.

ScottPJones replied:
I was in the process of changing it to `_nul` instead of `_null`, because it is the NUL character,
not a NULL.

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,208 @@
+### Keyword Arguments:
+* `accept_long_nul`   = `true` # Modified UTF-8 (`\\0` represented as `b\"\\xc0\\x80\"`)
+* `accept_surrogates` = `true` # `CESU-8`
+* `accept_long` = `false`
```
what's this one do then?

ScottPJones commented:
OK, not sure why it says the AppVeyor build was canceled... anyway, next round of changes done.

tkelman commented:
I cancelled it. There's an issue that will cause it to time out that we're in the process of fixing, and https://github.com/JuliaLang/julia/pull/11575/files#r31957096 is unaddressed.

ScottPJones commented:
Hopefully everything is all set now!

kmsquire commented:
I like the fact that `check_string` is now only two methods!

ScottPJones commented:
Anything else that needs doing on this?

tkelman commented:
Needs review (https://github.com/JuliaLang/julia/pull/11575#issuecomment-110017070) from @stevengj or @jiahao etc.

ScottPJones commented:
Just rebased with all of the changes to master to keep current... let me know if there is anything that needs changing...

tkelman commented:
There's generally no reason to rebase unless there's a conflict or you have code updates. Especially with multiple open PR's at once, it can make the appveyor backlog quite a bit worse if you rebase all of them in short succession.

ScottPJones commented:
The reason I rebased was that #11551 (UTF-16 bug fixes) depends on this, and #11607 (UTF-32 bug fixes) depends on #11551, and #11624 depends on #11607... and #11607 couldn't be automatically merged because of other changes to master... and #11607 is the branch I need for work.

ScottPJones commented:
Can anybody else review this and see about merging it?
It does not require any specific knowledge of Unicode (just look at the Wikipedia page: https://en.wikipedia.org/wiki/UTF-8 that has everything you need to know).
This is totally non-breaking (since until the follow-on PRs that actually fix the bugs in UTF-8, UTF-16, and UTF-32 conversions are merged, nothing in Base uses it yet).

nalimilan commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,208 @@
+    local byt::UInt8, ch::UInt32, surr::UInt32
+    flags::UInt = 0
+    totalchar = num2byte = num3byte = num4byte = 0
+    @inbounds while pos <= endpos
```
To what call to `getindex` does this `@inbounds` statement apply?

ScottPJones replied:
Doesn't it affect all of the `next()` calls?

nalimilan replied:
Yeah, but I'm not clear on whether `@inbounds` has an effect inside the functions you call. Only when they are inlined?

nalimilan commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,208 @@
+Validates and calculates number of characters in a UTF-16 or UTF-32 encoded vector/string
+
+### Input Arguments:
+* `dat`    Vector of `UInt16`, `UInt32`, or an `AbstractString`
```
This comment does not add anything since we have the signature. I'd say either remove it, or describe the expected *contents* of these arguments.

ScottPJones replied:
OK, good point... will do.

nalimilan commented:
Without saying anything about the actual checks, the code looks quite clean to me now. Also the fact that you add tests is great. But I would also add checks for a few valid strings, to make sure the functions don't start throwing an error for these cases in the future.

ScottPJones commented:
Yes, I'll add some tests for valid strings as well (those were part of the conversion tests, but since those were split off at reviewer request, they were no longer part of this PR).

ScottPJones commented:
OK, this now has tests for valid strings (not just a few!, but for all edge cases).
I also used the new generic function syntax, and combined the documentation.
Ready for more :tomato:s here!

stevengj commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,189 @@
+    local ch::UInt32
+    flags::UInt = 0
+    totalchar = num2byte = num3byte = num4byte = 0
+    @inbounds while pos <= endpos
```
Since `pos` and `endpos` are caller-provided, you need to check `pos >= start(dat)` and `endpos <= endof(dat)` before the loop body and throw a `BoundsError()`, in order to avoid crashing on invalid arguments due to `@inbounds`.

stevengj commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,189 @@
+### Throws:
+* `UnicodeError`
+"
+function check_string end
```
?

ScottPJones replied:
Just what is the ? about?

nalimilan replied:
I think @stevengj was wondering why you added this empty definition. I had the same reaction, though now I see that's because you merged the docs for both methods into one, and want to attach it to the function rather than to the methods. Makes sense to me (though currently the docs are not attached to anything of course.)

stevengj commented:
It seems weird to me to have a patch that introduces an undocumented internal function that isn't actually used anywhere (as of this PR).   Is the purpose of this just to shorten your other PR?  

Normally, I think, a PR should only introduce new undocumented internal functions if they are actually needed within the PR.

ScottPJones commented:
@stevengj I was told to break things up into smaller, more easily reviewed PRs.
As always, I've tried to comply as best as possible.
This is base functionality for a lot of bug fixing.
As far as being undocumented, once it is merged and in use, and any new issues that might be brought up have been dealt with, that is the time to make it a documented function.

I've added bounds checking now, in such a way that it should be able to totally skip the bounds checking for the cases where I use it (where the bounds are known in advance to be valid).
Please let me know if the technique of using `Val{true}` and `Val{false}`, which I borrowed from other places in Base, is correct (if not, then some other places in Base need to get changed as well).

nalimilan commented:
> Please let me know if the technique of using Val{true} and Val{false}, which I borrowed from other places in Base, is correct (if not, then some other places in Base need to get changed as well).

I don't think this use is justified here. Sounds like premature optimization. Checking bounds once before checking the whole string is most likely negligible give all the work you perform right after that. What matters is avoiding the check in each iteration of the loop.

tkelman commented:
I agree that using `Val{true}` and `Val{false}` for bounds-checking is not really warranted here.

@stevengj there are 3 other open PR's that build on this. I agree that introducing not-yet-used functionality is a bit of a strange thing to do, but the conversion PR's will need more work and given that we're close to moving to 0.4-pre I think it could be worth trying to get this infrastructure in place ahead of time and independently of the other changes that build upon it.

ScottPJones commented:
@nalimilan These functions will also be called frequently to convert very small strings - so I don't think it is "premature" optimization.  (Frequently, these can be called for many strings that are just 1 or less than 8 characters...)

@tkelman Is there some other technique that you would suggest, so that the conversion routines do not have to perform any unnecessary bounds checking, but that future callers would by default get bounds checking?

tkelman commented:
One option, not sure whether it's a good idea yet though, would be to call this `unsafe_checkstring` and do the bounds-checks separately in the callers.

tkelman commented:
Generally the time to document newly-added functions should be at the same time as adding them. That doesn't always get followed properly, e.g. for `llvmcall`. And we don't necessarily document all internal-use, non-exported functions.

ScottPJones commented:
@tkelman I'm fine with calling it unsafe_checkstring.  I do think all internal-use, non-exported functions *should* be (internally) documented, which is what I did with the functions in this and other PRs.
Do you want me to 1) rename it unsafe_checkstring 2) add a comment in the documentation warning just how it is unsafe (bounds checking) 3) remove the extra `bounds` argument 4) remove the extra methods that handle defaults and checking?

Thanks!

nalimilan commented:
+1 for calling it `unsafe_checkstring`, if you really want to skip bounds checking.

ScottPJones commented:
OK, folks!  It now has both `unsafe_checkstring` (which I will change my other PRs to call), and `checkstring` (which does bounds checking), and they each have documentation, `unsafe_checkstring` warns about the lack of bounds checking, and references `checkstring`, and `checkstring` references `unsafe_checkstring`, pointing out it's use if bounds are already known to be valid.
It does seem much cleaner this way, at least to me.  Thanks for bringing up the issue of bounds checking, @stevengj, and the rest of you for suggestions on how to have my cake and eat it too!
I've also added tests to make sure the bounds checking is working.

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,235 @@
+"
+function checkstring end
+
+# No need to check bounds if using defaults
```
how about just one method here to cut down on repetition
```
function checkstring(dat, endpos=endof(dat), pos=start(dat))
    pos < 1 && throw(BoundsError(dat, pos))
    (pos <= endpos <= endof(dat)) || throw(BoundsError(dat, endpos))
    unsafe_checkstring(dat, endpos, pos)
end
```

ScottPJones replied:
OK, sure.

ScottPJones replied:
For this function, at least, `pos` should be `startpos`... I'll change that.

kmsquire replied:
Before keyword arguments existed in Julia, something like the following was common (and is still perfectly valid):

```julia
function checkstring(dat, startpos, endpos)
   # code here
end

checkstring(dat) = checkstring(dat, start(dat), endof(dat))
checkstring(dat, endpos) = checkstring(dat, start(dat), endpos)

```

tkelman commented:
Slightly bikeshed-ish, but would `pos` be better-named `startpos`? And what's the reason for having `endpos` come before (start) `pos` in the input list, feels backwards.

ScottPJones commented:
About `endpos` before `pos`, that came about because in all the places I'd used it, I only needed to specify `endpos` (which used to be `length`), `pos` always started at `start(dat)`.
You're right though, I don't like the backwards order either...
I have a question... how can I collect and pass arbitrary keyword arguments from `checkstring` to `unsafe_checkstring`? And how can I skip a parameter? (i.e. if I change the order, but I still want to only pass the end position, and leave the start position defaulted?)  I tried `checkstring(dat,,len)` and
`checkstring(dat,_,len)`, but neither worked, and I can't find anything in the documentation...

tkelman commented:
For keyword arguments `f(; args...)` you can inspect and modify `args`. Don't think there's quite anything comparable for non-keyword positional arguments though.

ScottPJones commented:
OK, updated to change the order of the positions, `pos` -> `startpos` for the `checkstring` methods, now passing down the keyword arguments.

ScottPJones commented:
BTW, what's going on with AppVeyor and Travis-CI?  Is everything broken today?

tkelman commented:
Yes. Travis: #11691 AppVeyor: https://github.com/JuliaLang/julia/issues/11682#issuecomment-111713286

ScottPJones commented:
Is there anything else that is preventing this from being merged?  Thanks!

tkelman commented on base/utfcheck.jl:
```diff
@@ -0,0 +1,228 @@
+
+# Make sure that beginning and end positions are bounds checked
+function checkstring(dat, startpos = start(dat), endpos = endof(dat); kwargs...)
+    startpos < 1 && throw(BoundsError(dat, startpos))
```
should this be `startpos < start(dat)` ?

ScottPJones replied:
OK, good idea... I don't think it's necessary for any of the types that `unsafe_checkstring` *currently* accepts, but it's better to be future proof.

ScottPJones replied:
@tkelman, done!

tkelman commented:
I don't see any open unaddressed comments that github might eat, can you squash now?

ScottPJones commented:
@tkelman Done!
BTW, I looked in base for `BoundsError`, and if using 1 instead of start() is a problem (and I agree with you that it could be), then the majority of the bounds checks in base are incorrect...
Should I open an Issue on that? (with a PR to follow when I get a chance)?
Thanks!

nalimilan commented:
> BTW, I looked in base for BoundsError, and if using 1 instead of start() is a problem (and I agree with you that it could be), then the majority of the bounds checks in base are incorrect...
> Should I open an Issue on that? (with a PR to follow when I get a chance)?
> Thanks!

For `AbstractArrays` at least, we might consider it part of the interface that `1` is the first index. For other cases, maybe not. But I wouldn't bother to file an issue and change all calls unless we can find a concrete example where it matters: if you don't write tests with an actual type, things will get broken again without anybody noticing.

ScottPJones commented:
@nalimilan I found this in the code, that looks like it might be returning a tuple from `start`, however, being still fairly new to some of the details of julia, maybe I'm missing something:
```julia
# While the definitions for LinearFast are all simple enough to inline on their
# own, LinearSlow's CartesianRange is more complicated and requires explicit
# inlining.
start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
```

I'd just changed my code to use start(dat) instead of 1 at @tkelman's request, but I don't think it hurts anything even if it isn't strictly needed (start(dat) should in-line to 1 for my code, when used with `Vector`s, right?)

nalimilan commented:
Good catch indeed, things like `start(sparse([1 3]))` return a tuple. So if you find places where this case is not supported, and it should be, feel free to make a PR.

ScottPJones commented:
@nalimalan Unfortunately, in Base alone, there are many cases where people use 1 instead of start() when dealing with `AbstractArrays`, and also declare the index as being `Int`, which of course would break when getting a tuple...  Does `arr[ind]` work correctly, when `ind` is a `Tuple`? I would guess so,
but I want to be sure.  
I think for now, I need to raise an issue, which might need several PRs to fix (and not just from me, they'd better be done by the authors of the code that needs to be fixed).

ScottPJones commented:
Bump!  Is there anything still remaining preventing this from being merged?

nalimilan commented:
> Does arr[ind] work correctly, when ind is a Tuple? I would guess so,
> but I want to be sure.

Generally, no, you need to splat like this: `x[(1,2)...]`. But the result of `start(sparse([1 3]))` is not supposed to be used for indexing, only to call `next`. Looks like we would need `startind` or `startof` if we wanted to supported a non-integer start. But we can also decide that `1` should always be accepted by any indexable, even if it doesn't accept integer indexes.

Anyway -- we're quite far from the subject of the PR.

mbauman commented:
Iteration and indexing are two separate protocols. It used to be the case that, for arrays, they happened to be the same, and you could index into them with the iterator state. That is no longer true, as you have noticed.

Is there documentation that suggests that they must be linked? I think that would be the part that should change.

ScottPJones commented:
I've just added #11713 to continue discussing the issues raised here.
So, back to *this* PR :grinning: Is there *anything* else that needs to be done, or can somebody please push the button and merge this?   Thanks!

nalimilan commented:
Ah, something that @mbauman's comment on the other thread reminded me of: to check bounds in `checkstring`, you can call `checkbounds`, which is slightly shorter.

ScottPJones commented:
I'm already changing it - it doesn't really matter that it's shorter though, it is pretty important that it is correct! :grinning:  Thanks!

ScottPJones commented:
@nalimilan Turns out, there would still be problems, using `checkbounds`, so I think for now, this should just be left as is, and a separate PR done to deal with bounds checking issues throughout base.
So, anything left to do?

ScottPJones commented:
@mbauman Could you please take a look at my use of `next()`, `start()`, `endof()`, `pos`, and `endpos` in this code?  The back and forth review suggestions are driving me a bit crazy, and I'd like some sort of definitive answer as to what should be done.  I've just been following patterns I've seen elsewhere in the code, but if that code is not correct, then I'm lost...

mbauman commented:
Yes, I'm sorry, I was missing the context of this PR, and was answering all your questions (both here and in #11713) in the general case for all `AbstractArray`s.  Your usages of indexing and iteration over arrays here looks like its *entirely* restricted to `Vector`, and in that case, indexing and iteration state *happen* to be the same.  So everything here looks to be correct (albeit dependent upon the internal `Array` iteration implementation).  

I think that's just fine for internal code like this, but this is outside of my purview and I won't be helping to maintain this code directly.

ScottPJones commented:
@mbauman Thanks very much.  I'll be looking forward to your planned Interfaces manual page!

ScottPJones commented:
OK, I've changed this according to the very last round of suggestions in #11713.
Thanks everybody for your time and thorough reviews!
Anything left to fix?

tkelman commented:
Unless anyone objects strongly, I'll merge this in a day or two so we can continue on to the actual bug fixes. This part of the code looks pretty much done to me, so for the sake of future unicode bugfix backporting if nothing else let's get it on master.

ScottPJones commented:
:tada: Thank you very much, @tkelman! I'll go rebase the rest of my PRs now...
Thanks to everybody for the very thorough reviews, and all the time you spent on them.  I'm very impressed with and grateful for all the thought, great suggestions, and help from people!
